aesd-char-driver/main.c

```diff
@@ -0,0 +1,235 @@
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+#include "aesdchar.h"
+#include "linux/slab.h"
+#include "linux/string.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Aneesh Gurram"); /** TODO: fill in your name **/
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *my_dev;
+    PDEBUG("open");
+    my_dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = my_dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+     // As we didnt allocate anything I think nothing to be done.
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    struct aesd_buffer_entry *buffer = NULL;
+    ssize_t entry_offset_byte_return = 0, buffer_cnt = 0; 
+    if ( (filp == NULL) || (buf == NULL))
+    {
+        PDEBUG("ERROR: READ invalid ");
+        return -EINVAL;
+    }
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+    /**
+     * TODO: handle read
+     */
+     
+    if(mutex_lock_interruptible(&aesd_device.lock) !=0)
+    {
+    	PDEBUG("ERROR: MUTEX lock ");
+        return -EINTR;
+    }
+    buffer = aesd_circular_buffer_find_entry_offset_for_fpos(&aesd_device.buffer, *f_pos, &entry_offset_byte_return);
+    if(buffer == NULL)
+    {
+        *f_pos = 0;
+        goto error_handler;
+    }
+    buffer_cnt = buffer->size - entry_offset_byte_return;
+    buffer_cnt = (buffer_cnt > count) ? count : buffer_cnt;
+    *f_pos += buffer_cnt;
+    //copies data buf
+    if(copy_to_user(buf, buffer->buffptr+entry_offset_byte_return, buffer_cnt))
+    {
+        retval = -EFAULT;
+        goto error_handler;
+    }
+
+    retval = buffer_cnt;
+error_handler:
+    mutex_unlock(&aesd_device.lock);
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    struct aesd_dev *my_dev = NULL;  
+    const char *free_ptr = NULL;
+    if ( (filp == NULL) || (buf == NULL))
+    {
+        PDEBUG("ERROR: WRITE invalid ");
+        return -EINVAL;
+    }
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+    /**
+     * TODO: handle write
+     */
+    my_dev = filp->private_data;
+    if(!my_dev)
+    {
+        return -EFAULT;
+    }
+    
+    if(!(my_dev->entry.size))
```

Comment:
@aneesh1298 you either need to use the lock you have for the aesd_dev structure on accesses to this entry or you need another lock specific to this entry.  Otherwise multiple processes writing at the same time could cause issues with this access.